File: App/android/app/src/main/java/com/guru/batteryoptimizer/GuruAPIs.java
package com.guru.batteryoptimizer;

import android.app.Activity;
import android.os.Vibrator;
import android.os.VibratorManager;
import android.webkit.JavascriptInterface;
import android.view.View;
import android.view.HapticFeedbackConstants;
import android.content.Intent;
import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import org.json.JSONArray;
import org.json.JSONObject;
import java.io.File;
import java.io.InputStream;
import java.io.FileOutputStream;
import java.net.URL;
import java.net.HttpURLConnection;
import java.util.List;
import com.topjohnwu.superuser.Shell;

public class GuruAPIs {
    private Activity activity;
    static {
        Shell.setDefaultBuilder(Shell.Builder.create()
            .setFlags(Shell.FLAG_MOUNT_MASTER).setTimeout(10));
    }
    public GuruAPIs(Activity activity) {
        this.activity = activity;
    }
    // Haptic Vibrator Helper
    private Vibrator getHapticVibrator(Context context) {
        VibratorManager vm = context.getSystemService(VibratorManager.class);
        return vm != null ? vm.getDefaultVibrator() : null;
    }
    // App Updater Helper
    private void downloadFile(String urlStr, File outFile) throws Exception {
        URL url = new URL(urlStr);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setConnectTimeout(10000);
        conn.setReadTimeout(10000);
        conn.connect();
        try (InputStream in = conn.getInputStream();
            FileOutputStream out = new FileOutputStream(outFile)) {
            byte[] buffer = new byte[8192];
            int len;
            while ((len = in.read(buffer)) > 0) {
                out.write(buffer, 0, len);
            }
        }
    }
    // ============ BASIC APIs ============
    // = guru.openURL() =
    @JavascriptInterface
    public void openURL(String url) {
        try {
            Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            intent.addFlags(Intent.FLAG_ACTIVITY_EXCLUDE_FROM_RECENTS);
            activity.startActivity(intent);
        } catch (Exception ignored) {}
    }
    // = guru.closeApp() =
    @JavascriptInterface
    public void closeApp() {
        try {
            closeShell();
            activity.finish();
            android.os.Process.killProcess(android.os.Process.myPid());
            System.exit(0);
        } catch (Exception ignored) {}
    }
     // = guru.hideSplash() =
    @JavascriptInterface
    public void hideSplash() {
        activity.runOnUiThread(() -> {
            if (activity instanceof MainActivity) {
                ((MainActivity) activity).hideNativeSplash();
            }
        });
    }
    // = guru.getAppInfo()
    @JavascriptInterface
    public String getAppInfo() {
        try {
            PackageInfo pInfo = activity.getPackageManager().getPackageInfo(activity.getPackageName(), 0);
            JSONObject result = new JSONObject();
            result.put("success", true);
            long versionCode = pInfo.getLongVersionCode();
            result.put("versionCode", versionCode);
            result.put("versionName", pInfo.versionName);
            result.put("packageName", activity.getPackageName());
            result.put("updateUrl", "https://raw.githubusercontent.com/EliezerB03/Battery-GURU-Optimizer/master/App/updater.json");
            return result.toString();
        } catch (Exception e) {
            try {
                JSONObject error = new JSONObject();
                error.put("success", false);
                error.put("error", e.getMessage());
                return error.toString();
            } catch (Exception je) {
                return "{\"success\":false,\"error\":\"Unknown error\"}";
            }
        }
    }
    // ============== HAPTIC VIBRATIONS APIs ==============
    // = guru.hapticToggle()
    @JavascriptInterface
    public void hapticToggle() {
        activity.runOnUiThread(() -> {
            try {
                View rootView = activity.getWindow().getDecorView().getRootView();
                if (rootView != null) {
                    int hapticType;
                    if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
                        hapticType = HapticFeedbackConstants.TOGGLE_ON;
                    } else {
                        hapticType = HapticFeedbackConstants.CONFIRM;
                    }
                    rootView.performHapticFeedback(
                        hapticType, HapticFeedbackConstants.FLAG_IGNORE_VIEW_SETTING
                    );
                }
            } catch (Exception ignored) {}
        });
    }
    // = guru.hapticSlider()
    @JavascriptInterface
    public void hapticSlider() {
        activity.runOnUiThread(() -> {
            try {
                View rootView = activity.getWindow().getDecorView().getRootView();
                if (rootView != null) {
                    rootView.performHapticFeedback(
                        HapticFeedbackConstants.CLOCK_TICK, HapticFeedbackConstants.FLAG_IGNORE_VIEW_SETTING
                    );
                }
            } catch (Exception ignored) {}
        });
    }
    // ============== ROOT APIs ==============
    // = guru.isRoot() =
    @JavascriptInterface
    public String isRoot() {
        try {
            closeShell();
            Shell shell = Shell.getShell();
            boolean hasRoot = shell.isRoot();
            JSONObject result = new JSONObject();
            result.put("hasRoot", hasRoot);
            result.put("success", true);
            return result.toString();
        } catch (Exception e) {
            try {
                JSONObject error = new JSONObject();
                error.put("success", false);
                error.put("hasRoot", false);
                error.put("error", e.getMessage());
                return error.toString();
            } catch (Exception je) {
                return "{\"success\":false,\"hasRoot\":false,\"error\":\"Unknown error\"}";
            }
        }
    }
    // = guru.getDeviceInfo() =
    @JavascriptInterface
    public String getDeviceInfo() {
        try {
            List<String> out = Shell.cmd("getprop ro.boot.em.model").exec().getOut();
            String model = (out != null && !out.isEmpty()) ? out.get(0).trim() : "";
            JSONObject result = new JSONObject();
            result.put("success", true);
            result.put("model", model);
            result.put("raw", out); 
            return result.toString();
        } catch (Exception e) {
            try {
                JSONObject error = new JSONObject();
                error.put("success", false);
                error.put("error", e.getMessage());
                return error.toString();
            } catch (Exception je) {
                return "{\"success\":false,\"error\":\"Unknown error\"}";
            }
        }
    }
    // = guru.closeShell() =
    public void closeShell() {
        try {
            Shell shell = Shell.getCachedShell();
            if (shell != null && shell.isAlive()) {
                shell.close();
            }
        } catch (Exception e) {}
    }
    // = guru.installUpdate() =
    @JavascriptInterface
    public void installUpdate(String url) {
        new Thread(() -> {
            try {
                File apkFile = new File(activity.getCacheDir(), "update.apk");
                downloadFile(url, apkFile);
                Shell.Result result = Shell.cmd("pm install -r " + apkFile.getAbsolutePath()).exec();
                activity.runOnUiThread(() -> {
                    activity.finishAffinity();
                    System.exit(0);
                });
            } catch (Exception e) {}
        }).start();
    }
    // = guru.exec() =
    @JavascriptInterface
    public String exec(String command) {
        try {
            Shell.Result result = Shell.cmd(command).exec();
            if (!result.isSuccess()) {
                return "";
            }
            StringBuilder output = new StringBuilder();
            for (String line : result.getOut()) {
                if (output.length() > 0) {
                    output.append("\n");
                }
                output.append(line);
            }
            return output.toString();
        } catch (Exception e) {
            return "";
        }
    }
}

File: App/android/guru/src/main/java/com/guru/batteryoptimizer/utils/GuruAPIs.kt
package com.guru.batteryoptimizer.utils

import android.content.Context
import android.content.Intent
import android.os.Handler
import android.os.Looper
import com.topjohnwu.superuser.Shell
import kotlinx.coroutines.*
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.io.File
import java.io.FileOutputStream
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.atomic.AtomicBoolean
import org.json.JSONObject

class GuruAPIs(private val context: Context) {
    // ======= Objet Variables =======
    companion object {
        init {
            Shell.setDefaultBuilder(
                Shell.Builder.create()
                    .setFlags(Shell.FLAG_MOUNT_MASTER)
                    .setTimeout(10)
            )
        }
    }
    private val mainHandler = Handler(Looper.getMainLooper())
    private var currentActivity: java.lang.ref.WeakReference<android.app.Activity>? = null
    private fun runOnUiThread(block: () -> Unit) = mainHandler.post(block)

    // ======= Download State =======
    var isDownloading: Boolean = false
        private set
    var downloadProgress: Int = 0
        private set
    var downloadCompletedWhileDetached: Boolean = false
    private val downloadCancelFlag = AtomicBoolean(false)
    private var onProgressCallback: ((Int) -> Unit)? = null
    private var onCompleteCallback: (() -> Unit)? = null
    private var onCancelCallback: (() -> Unit)? = null
    private var onErrorCallback: (() -> Unit)? = null
    fun reattachDownloadCallbacks(
        onProgress: ((Int) -> Unit)?,
        onComplete: () -> Unit,
        onCancel: () -> Unit,
        onError: () -> Unit
    ) {
        onProgressCallback = onProgress
        onCompleteCallback = onComplete
        onCancelCallback   = onCancel
        onErrorCallback    = onError
    }
    fun detachDownloadCallbacks() {
        onProgressCallback = null
        onCompleteCallback = null
        onCancelCallback   = null
        onErrorCallback    = null
    }
    fun cancelActiveDownload() {
        downloadCancelFlag.set(true)
    }

    // ======= Reset State =======
    var isResetting: Boolean = false
        private set
    
    // ======= Optimizing Apps State =======
    var isAppOptimizing: Boolean = false
        private set
    var onOptimizeProgress: ((Int) -> Unit)? = null
    var onOptimizeComplete: (() -> Unit)? = null
    var onOptimizeCancel: (() -> Unit)? = null
    private val optimizeCancelFlag = AtomicBoolean(false)
    fun cancelActiveOptimize() {
        optimizeCancelFlag.set(true)
    }

    // ======= Helpers =======
    private fun errorJson(e: Exception) = runCatching {
        JSONObject().apply {
            put("success", false)
            put("error", e.message)
        }.toString()
    }.getOrElse { """{"success":false,"error":"Unknown error"}""" }

    // ======= Download File Helper =======
    private fun downloadFile(
        urlStr: String,
        outFile: File,
        onProgress: ((Int) -> Unit)? = null,
        isCancelled: () -> Boolean = { false }
    ) {
        val conn = (URL(urlStr).openConnection() as HttpURLConnection).apply {
            connectTimeout = 10_000
            readTimeout    = 10_000
            connect()
        }
        val totalBytes = conn.contentLengthLong
        var downloaded = 0L
        conn.inputStream.use { input ->
            FileOutputStream(outFile).use { output ->
                val buffer = ByteArray(8192)
                var bytes: Int
                while (input.read(buffer).also { bytes = it } != -1) {
                    if (isCancelled()) {
                        conn.disconnect()
                        throw CancellationException("Download cancelled")
                    }
                    output.write(buffer, 0, bytes)
                    downloaded += bytes
                    if (totalBytes > 0) {
                        val percent = ((downloaded * 100) / totalBytes).toInt()
                        onProgress?.invoke(percent)
                    }
                }
            }
        }
        conn.disconnect()
    }
    private fun downloadFileWithRetry(
        urlStr: String,
        outFile: File,
        onProgress: ((Int) -> Unit)? = null,
        isCancelled: () -> Boolean = { false }
    ) {
        val retryDeadline   = System.currentTimeMillis() + 10_000L
        val retryIntervalMs = 500L
        var lastException: Exception? = null
        var attempted = false
        while (!attempted || System.currentTimeMillis() < retryDeadline) {
            if (isCancelled()) throw CancellationException("Download cancelled")
            try {
                downloadFile(urlStr, outFile, onProgress, isCancelled)
                return
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                lastException = e
                attempted = true
                if (System.currentTimeMillis() < retryDeadline) {
                    Thread.sleep(retryIntervalMs)
                }
            }
        }
        throw lastException ?: Exception("Download failed")
    }

    // ============ APIs (ROOT) ============
    @Volatile var cancelRunningOperations: Boolean = false
    fun isRoot(): String = runCatching {
        val cached = Shell.getCachedShell()
        if (cached != null && cached.isAlive) {
            val hasRoot = runCatching {
                Shell.Builder.create()
                    .setFlags(Shell.FLAG_MOUNT_MASTER)
                    .setTimeout(5)
                    .build("su")
                    .use { it.isRoot }
            }.getOrElse { false }
            if (!hasRoot) closeShell()
            JSONObject().apply {
                put("success", true)
                put("hasRoot", hasRoot)
            }.toString()
        } else {
            closeShell()
            val hasRoot = Shell.getShell().isRoot
            JSONObject().apply {
                put("success", true)
                put("hasRoot", hasRoot)
            }.toString()
        }
    }.getOrElse { e ->
        runCatching {
            JSONObject().apply {
                put("success", false)
                put("hasRoot", false)
                put("error", (e as Exception).message)
            }.toString()
        }.getOrElse { """{"success":false,"hasRoot":false,"error":"Unknown error"}""" }
    }
    // getDeviceInfo
    fun getDeviceInfo(): String = runCatching {
        val model = Shell.cmd("getprop ro.boot.em.model").exec().out
            .firstOrNull()?.trim().orEmpty()
        JSONObject().apply {
            put("success", true)
            put("model",   model)
        }.toString()
    }.getOrElse { e -> errorJson(e as Exception) }
    // downloadUpdate
    fun downloadUpdate(
        url: String,
        onProgress: ((Int) -> Unit)? = null,
        onComplete: () -> Unit,
        onCancel: () -> Unit,
        onError: () -> Unit
    ) {
        downloadCancelFlag.set(false)
        isDownloading = true
        downloadProgress = 0
        onProgressCallback = onProgress
        onCompleteCallback = onComplete
        onCancelCallback   = onCancel
        onErrorCallback    = onError
        val apkFile = File(context.cacheDir, "update.apk")
        Thread {
            runCatching {
                downloadFileWithRetry(url, apkFile, onProgress = { percent ->
                    downloadProgress = percent
                    runOnUiThread { onProgressCallback?.invoke(percent) }
                }, isCancelled = { downloadCancelFlag.get() || cancelRunningOperations })
                if (cancelRunningOperations) {
                    isDownloading = false
                    runCatching { apkFile.delete() }
                    return@runCatching
                }
                isDownloading = false
                if (onCompleteCallback != null) {
                    runOnUiThread { onCompleteCallback?.invoke() }
                } else {
                    downloadCompletedWhileDetached = true
                }
            }.onFailure { e ->
                isDownloading = false
                runCatching { apkFile.delete() }
                if (!cancelRunningOperations) runOnUiThread {
                    if (e is CancellationException) onCancelCallback?.invoke()
                    else onErrorCallback?.invoke()
                }
            }
        }.start()
    }
    // installCachedUpdate
    fun installCachedUpdate(onError: () -> Unit) {
        Thread {
            if (cancelRunningOperations) return@Thread
            runCatching {
                Thread.sleep(100)
                if (cancelRunningOperations) return@Thread
                Thread.sleep(1500)
                val apkFile = File(context.cacheDir, "update.apk")
                if (cancelRunningOperations) return@Thread
                if (!apkFile.exists()) throw Exception("APK not found in cache")
                Shell.cmd("pm install -r ${apkFile.absolutePath}").exec()
                runCatching { apkFile.delete() }
                runOnUiThread {
                    android.os.Process.killProcess(android.os.Process.myPid())
                    System.exit(0)
                }
            }.onFailure {
                runOnUiThread { onError() }
            }
        }.start()
    }
    // resetToDefault
    fun resetToDefault() {
        if (isResetting) return
        isResetting = true
        Thread {
            if (cancelRunningOperations) { isResetting = false; return@Thread }
            runCatching {
                if (cancelRunningOperations) { isResetting = false; return@Thread }
                val pkg  = "com.guru.batteryoptimizer"
                val base = "/data/data/$pkg"
                val destPath = "$base/guru.sh"
                val escaped = GuruFunctions.Scripts.GURU_SCRIPT.replace("'", "'\\''")
                Thread.sleep(1600)
                if (cancelRunningOperations) { isResetting = false; return@Thread }
                exec("printf '%s' '$escaped' > $destPath")
                exec("chmod 755 $destPath")
                exec("sh $destPath")
                exec("find $base -mindepth 1 -maxdepth 1 -exec rm -rf {} +")
                Thread.sleep(100)
                runOnUiThread { closeApp() }
            }
        }.start()
    }
    // optimizeAllApps
    fun optimizeAllApps() {
        if (isAppOptimizing) return
        isAppOptimizing = true
        optimizeCancelFlag.set(false)
        Thread {
            if (cancelRunningOperations || optimizeCancelFlag.get()) { isAppOptimizing = false; return@Thread }
            runCatching {
                if (cancelRunningOperations || optimizeCancelFlag.get()) { isAppOptimizing = false; return@Thread }
                Thread.sleep(500)
                if (cancelRunningOperations || optimizeCancelFlag.get()) {
                    isAppOptimizing = false
                    runOnUiThread { onOptimizeCancel?.invoke() }
                    return@Thread
                }
                val packages = exec("pm list packages | cut -d: -f2").lines().map {it.trim()}.filter {it.isNotEmpty()}
                val total = packages.size
                var done = 0
                for (pkg in packages) {
                    if (cancelRunningOperations || optimizeCancelFlag.get()) {
                        isAppOptimizing = false
                        runOnUiThread { onOptimizeCancel?.invoke() }
                        return@Thread
                    }
                    exec("cmd package compile -m speed-profile \"$pkg\" > /dev/null 2>&1")
                    done++
                    val percent = if (total > 0) ((done * 100) / total) else 0
                    runOnUiThread { onOptimizeProgress?.invoke(percent) }
                }
                Thread.sleep(100)
                isAppOptimizing = false
                runOnUiThread { onOptimizeComplete?.invoke() }
            }
        }.start()
    }
    // execSafe (exec function)
    private val execScope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private val pendingJobs = mutableMapOf<String, Deferred<String>>()
    private val execMutex = Mutex()
    suspend fun execSafe(key: String, command: String): String {
        val deferred = execMutex.withLock {
            pendingJobs[key]?.cancel()
            val d = execScope.async { runCatching { exec(command) }.getOrElse { "" } }
            pendingJobs[key] = d; d
        }
        return try {
            deferred.await()
        } catch (_: CancellationException) {
            ""
        } finally {
            execMutex.withLock { if (pendingJobs[key] === deferred) pendingJobs.remove(key) }
        }
    }
    // exec (core function)
    fun exec(command: String): String = runCatching {
        val result = Shell.cmd(command).exec()
        if (!result.isSuccess) return ""
        result.out.joinToString("\n")
    }.getOrElse { "" }
    
    // ============ APIs (NON ROOT) ============
    // closeApp
    fun closeApp() {
        runCatching {
            execScope.cancel()
            closeShell()
        }
        Handler(Looper.getMainLooper()).post {
            currentActivity?.get()?.finish()
            Handler(Looper.getMainLooper()).postDelayed({ android.os.Process.killProcess(android.os.Process.myPid()) }, 250)
        }
    }
    fun attachActivity(activity: android.app.Activity) {currentActivity = java.lang.ref.WeakReference(activity)}
    // restartApp
    fun restartApp() {
        Handler(Looper.getMainLooper()).post {
            val intent = context.packageManager.getLaunchIntentForPackage(context.packageName) ?: return@post
            val mainIntent = Intent.makeRestartActivityTask(intent.component)
            context.startActivity(mainIntent)
            Runtime.getRuntime().exit(0)
        }
    }
    // closeShell
    fun closeShell() { runCatching { Shell.getCachedShell()?.takeIf { it.isAlive }?.close() } }
    // getAppInfo
    fun getAppInfo(): String = runCatching {
        val pInfo = context.packageManager.getPackageInfo(context.packageName, 0)
        JSONObject().apply {
            put("success",     true)
            put("versionCode", pInfo.longVersionCode)
            put("versionName", pInfo.versionName)
            put("packageName", context.packageName)
            put("updateUrl",   "https://raw.githubusercontent.com/EliezerB03/Battery-GURU-Optimizer/master/App/updater.json")
        }.toString()
    }.getOrElse { e -> errorJson(e as Exception) }
}